pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -255,12 +256,65 @@ func configFromEnv() (cfg Config, ok bool) {
 	return
 }
 
+// Allows Setting up of openstack credentials by reading clouds.yaml
+//
+// Looks for clouds.yaml file in the following paths in order:
+// 1. OS_CLIENT_CONFIG_FILE
+// 2. Current directory.
+// 3. unix-specific user_config_dir (~/.config/openstack/clouds.yaml)
+// 4. unix-specific site_config_dir (/etc/openstack/clouds.yaml)
+func configFromYAML() (cfg Config, err error) {
+	clientOpts := new(clientconfig.ClientOpts)
+	cloud, err := clientconfig.GetCloudFromYAML(clientOpts)
+
+	// Exit as a failure if unable to read clouds.yaml
+	if err != nil {
+		fmt.Println(err)
+		return cfg, err
+	}
+
+	cfg.Global.AuthURL = cloud.AuthInfo.AuthURL
+	cfg.Global.Username = cloud.AuthInfo.Username
+	cfg.Global.Password = cloud.AuthInfo.Password
+	cfg.Global.Region = cloud.RegionName
+	cfg.Global.UserID = cloud.AuthInfo.UserID
+	cfg.Global.TenantID = cloud.AuthInfo.ProjectID
+	cfg.Global.TenantName = cloud.AuthInfo.ProjectName
+	cfg.Global.DomainID = cloud.AuthInfo.UserDomainID
+	cfg.Global.DomainName = cloud.AuthInfo.UserDomainName
+	cfg.Global.CAFile = cloud.CACertFile
+
+	// Retaining support for trusts via environment variables
+	// due to lack of support by GetCloudFromYAML(gohpercloud)
+	cfg.Global.TrustID = os.Getenv("OS_TRUST_ID")
+
+	// Supporting injecting passwords via env variables
+	if cfg.Global.Password == "" {
+		cfg.Global.Password = os.Getenv("OS_PASSWORD")
+	}
+
+	cfg.Metadata.SearchOrder = fmt.Sprintf("%s,%s", configDriveID, metadataID)
+	cfg.BlockStorage.BSVersion = "auto"
+	cfg.Networking.IPv6SupportDisabled = false
+	cfg.Networking.PublicNetworkName = "public"
+
+	err = nil
+	return
+}
+
 func ReadConfig(config io.Reader) (Config, error) {
 	if config == nil {
 		return Config{}, fmt.Errorf("no OpenStack cloud provider config file given")
 	}
+	var cfg Config
 
-	cfg, _ := configFromEnv()
+	// Prioritizes reading clouds.yaml over envorinment variables
+	cfg, err := configFromYAML()
+	if err != nil && err.Error() == "unable to load clouds.yaml: no clouds.yaml file found" {
+		cfg, _ = configFromEnv()
```

Comment:
To properly keep backwards compatibility, I think we should first attempt to load configs using the existing method, warn the user that this behavior will change (only if the configs were set using the old method), then attempt to load them using `clouds.yaml`.

Reply:
I'm not sure this is an issue. According to the unit tests, the app has full backwards compatibility. I personally think that clouds.yaml should take priority, since it is generated directly by openstack, is a more reliable source of the data we need, and is ultimately the configuration method we want to support moving forward.

Reply:
Agreed that `clouds.yaml` is a better source for this config. However, there may be deployments currently using the "old method". If one of these deployments updates the provider and has a valid `clouds.yaml` in a valid path, the deployment may end up using that instead of the currently known method. We don't want to change the current order without first warning OPs that this change is coming.